kotlin/viewshed-geoprocessing/src/main/java/com/esri/arcgisruntime/sample/viewshedgeoprocessing/MainActivity.kt

```diff
@@ -0,0 +1,218 @@
+    val newFeature = featureCollectionTable.createFeature().apply {
+      geometry = point
+    }
+
+    // add newFeature and call perform Geoprocessing on done loading
+    featureCollectionTable.addFeatureAsync(newFeature)
+    featureCollectionTable.addDoneLoadingListener {
+      if (featureCollectionTable.loadStatus == LoadStatus.LOADED) {
+        performGeoprocessing(featureCollectionTable)
+      }
+    }
+  }
+
+  /**
+   * Creates a GeoprocessingJob from the GeoprocessingTask. Displays the resulting viewshed on the map.
+   *
+   * @param featureCollectionTable containing the observation point.
+   */
+  private fun performGeoprocessing(featureCollectionTable: FeatureCollectionTable) {
+
+    // geoprocessing parameters
+    val parameterFuture: ListenableFuture<GeoprocessingParameters> =
+      geoprocessingTask.createDefaultParametersAsync()
+    parameterFuture.addDoneListener {
+      try {
+        val parameters = parameterFuture.get().apply {
+          processSpatialReference = featureCollectionTable.spatialReference
+          outputSpatialReference = featureCollectionTable.spatialReference
+
+          // use the feature collection table to create the required GeoprocessingFeatures input
+          inputs["Input_Observation_Point"] = GeoprocessingFeatures(featureCollectionTable)
+        }
+
+        // initialize job from geoprocessingTask
+        geoprocessingJob = geoprocessingTask.createJob(parameters)
+
+        // start the job
+        geoprocessingJob?.start()
+
+        // listen for job success
+        geoprocessingJob?.addJobDoneListener {
+          if (geoprocessingJob?.status == Job.Status.SUCCEEDED) {
+            // get the viewshed from geoprocessingResult
+            (geoprocessingJob?.result?.outputs?.get("Viewshed_Result") as? GeoprocessingFeatures)?.let { viewshedResult ->
+              // for each feature in the result
+              for (feature in viewshedResult.features) {
+                // add the feature as a graphic
+                resultGraphicsOverlay.graphics.add(Graphic(feature.geometry))
+              }
+            }
+          } else {
+            Toast.makeText(
+              applicationContext,
+              "Geoprocessing result failed!",
+              Toast.LENGTH_LONG
+            ).show()
+          }
+        }
+      } catch (e: Exception) {
+        when (e) {
+          is InterruptedException, is ExecutionException -> ("Error getting geoprocessing result: " + e.message).also {
+            Toast.makeText(this, it, Toast.LENGTH_LONG).show()
+            Log.e(TAG, it)
+          }
+          else -> throw e
+        }
+      }
+    }
+  }
+
+  override fun onResume() {
+    super.onResume()
+    mapView.resume()
+  }
+
+  override fun onPause() {
+    mapView.pause()
+    super.onPause()
+  }
+
+  override fun onDestroy() {
+    mapView.dispose()
+    super.onDestroy()
+  }
+
+  companion object {
+    private val TAG: String = this::class.java.simpleName
```

Comment:
I've seen these in a few places across our samples now: most commonly at the top where the member variables are assigned, I think this is better placed up there. The companion object notation is new to me, I've now seen it a few times. I've just been using `private val TAG: String = MainActivity::class.java.simpleName`. Curious as to what is the advantage of companion object over that?

Reply:
I think in this case just having `private val TAG` at the top, not in a companion object, makes sense. I guess because this is the equivalent of Java's final static var, this might have been auto generated? It makes a bit more sense in another sample where I have other stuff inside the companion object. 

So I guess the advantage is kind of irrelevant in samples, but in an app it might make sense to be able to access the TAG statically...maybe. I'll move this out of the companion for this anyway because just having a private val looks cleaner

Reply:
sounds good to me, thanks for explaining!